File: server/tests/unit/auth/utils.test.ts
import {
  hashPassword,
  comparePassword,
  generateToken,
  verifyToken,
  isValidEmail,
  isValidUsername,
  isValidPassword,
} from '../../../src/auth/utils';
import jwt from 'jsonwebtoken';

describe('Auth Utils', () => {
  describe('hashPassword', () => {
    it('should hash a password', async () => {
      const password = 'MyPassword123';
      const hash = await hashPassword(password);

      expect(hash).toBeDefined();
      expect(hash).not.toBe(password);
      expect(hash.length).toBeGreaterThan(0);
    });

    it('should generate different hashes for same password', async () => {
      const password = 'MyPassword123';
      const hash1 = await hashPassword(password);
      const hash2 = await hashPassword(password);

      expect(hash1).not.toBe(hash2); // bcrypt uses random salt
    });
  });

  describe('comparePassword', () => {
    it('should return true for matching password', async () => {
      const password = 'MyPassword123';
      const hash = await hashPassword(password);

      const result = await comparePassword(password, hash);

      expect(result).toBe(true);
    });

    it('should return false for non-matching password', async () => {
      const password = 'MyPassword123';
      const wrongPassword = 'WrongPassword456';
      const hash = await hashPassword(password);

      const result = await comparePassword(wrongPassword, hash);

      expect(result).toBe(false);
    });

    it('should return false for empty password', async () => {
      const password = 'MyPassword123';
      const hash = await hashPassword(password);

      const result = await comparePassword('', hash);

      expect(result).toBe(false);
    });
  });

  describe('generateToken', () => {
    it('should generate a valid JWT token', () => {
      const userId = 'user-123';
      const username = 'testuser';

      const token = generateToken(userId, username);

      expect(token).toBeDefined();
      expect(typeof token).toBe('string');
      expect(token.split('.').length).toBe(3); // JWT has 3 parts
    });

    it('should include userId and username in token payload', () => {
      const userId = 'user-123';
      const username = 'testuser';

      const token = generateToken(userId, username);
      const decoded = jwt.decode(token) as any;

      expect(decoded.userId).toBe(userId);
      expect(decoded.username).toBe(username);
    });

    it('should generate different tokens for different users', () => {
      const token1 = generateToken('user-1', 'user1');
      const token2 = generateToken('user-2', 'user2');

      expect(token1).not.toBe(token2);
    });
  });

  describe('verifyToken', () => {
    it('should verify and decode a valid token', () => {
      const userId = 'user-123';
      const username = 'testuser';
      const token = generateToken(userId, username);

      const decoded = verifyToken(token);

      expect(decoded.userId).toBe(userId);
      expect(decoded.username).toBe(username);
    });

    it('should throw error for invalid token', () => {
      const invalidToken = 'invalid.token.here';

      expect(() => verifyToken(invalidToken)).toThrow('Invalid or expired token');
    });

    it('should throw error for malformed token', () => {
      const malformedToken = 'not-a-jwt-token';

      expect(() => verifyToken(malformedToken)).toThrow('Invalid or expired token');
    });

    it('should throw error for empty token', () => {
      expect(() => verifyToken('')).toThrow('Invalid or expired token');
    });
  });

  describe('isValidEmail', () => {
    it('should return true for valid emails', () => {
      expect(isValidEmail('user@example.com')).toBe(true);
      expect(isValidEmail('test.user@example.com')).toBe(true);
      expect(isValidEmail('user+tag@example.co.uk')).toBe(true);
      expect(isValidEmail('user123@test-domain.com')).toBe(true);
      expect(isValidEmail('a@b.co')).toBe(true);
    });

    it('should return false for invalid emails', () => {
      expect(isValidEmail('invalid')).toBe(false);
      expect(isValidEmail('invalid@')).toBe(false);
      expect(isValidEmail('@example.com')).toBe(false);
      expect(isValidEmail('user@.com')).toBe(false);
      expect(isValidEmail('user@example')).toBe(false);
      expect(isValidEmail('user @example.com')).toBe(false);
      expect(isValidEmail('user@exam ple.com')).toBe(false);
      expect(isValidEmail('')).toBe(false);
    });
  });

  describe('isValidUsername', () => {
    it('should return true for valid usernames', () => {
      expect(isValidUsername('user123')).toBe(true);
      expect(isValidUsername('test_user')).toBe(true);
      expect(isValidUsername('Player_123')).toBe(true);
      expect(isValidUsername('abc')).toBe(true); // minimum 3 chars
      expect(isValidUsername('a'.repeat(50))).toBe(true); // maximum 50 chars
    });

    it('should return false for invalid usernames', () => {
      expect(isValidUsername('ab')).toBe(false); // too short
      expect(isValidUsername('a'.repeat(51))).toBe(false); // too long
      expect(isValidUsername('user-name')).toBe(false); // dash not allowed
      expect(isValidUsername('user.name')).toBe(false); // dot not allowed
      expect(isValidUsername('user name')).toBe(false); // space not allowed
      expect(isValidUsername('user@name')).toBe(false); // special char
      expect(isValidUsername('user#123')).toBe(false); // special char
      expect(isValidUsername('')).toBe(false); // empty
    });
  });

  describe('isValidPassword', () => {
    it('should return true for valid passwords', () => {
      expect(isValidPassword('Password123')).toBe(true);
      expect(isValidPassword('Abcd1234')).toBe(true);
      expect(isValidPassword('MyP@ssw0rd')).toBe(true);
      expect(isValidPassword('Test1234')).toBe(true);
    });

    it('should return false for passwords without uppercase', () => {
      expect(isValidPassword('password123')).toBe(false);
      expect(isValidPassword('test1234')).toBe(false);
    });

    it('should return false for passwords without lowercase', () => {
      expect(isValidPassword('PASSWORD123')).toBe(false);
      expect(isValidPassword('TEST1234')).toBe(false);
    });

    it('should return false for passwords without numbers', () => {
      expect(isValidPassword('Password')).toBe(false);
      expect(isValidPassword('TestPass')).toBe(false);
    });

    it('should return false for passwords shorter than 8 characters', () => {
      expect(isValidPassword('Pass1')).toBe(false);
      expect(isValidPassword('Ab1')).toBe(false);
      expect(isValidPassword('')).toBe(false);
    });

    it('should accept passwords with special characters', () => {
      expect(isValidPassword('Pass@123')).toBe(true);
      expect(isValidPassword('Test!234')).toBe(true);
      expect(isValidPassword('My#Pass1')).toBe(true);
    });
  });
});

File: go-server/internal/auth/utils_test.go
package auth

import (
	"strings"
	"testing"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

func TestHashPassword(t *testing.T) {
	tests := []struct {
		name        string
		password    string
		expectError bool
	}{
		{
			name:        "valid password",
			password:    "MyPassword123",
			expectError: false,
		},
		{
			name:        "short password",
			password:    "Pass1",
			expectError: false,
		},
		{
			name:        "long password (within bcrypt limit)",
			password:    strings.Repeat("a", 70),
			expectError: false,
		},
		{
			name:        "password exceeds bcrypt limit",
			password:    strings.Repeat("a", 100),
			expectError: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			hash, err := HashPassword(tt.password)

			if tt.expectError {
				if err == nil {
					t.Error("HashPassword() should return error for password exceeding bcrypt limit")
				}
				return
			}

			if err != nil {
				t.Fatalf("HashPassword() error = %v", err)
			}

			if hash == "" {
				t.Error("HashPassword() returned empty string")
			}

			if hash == tt.password {
				t.Error("HashPassword() returned password unchanged")
			}

			if len(hash) == 0 {
				t.Error("HashPassword() returned zero length hash")
			}
		})
	}
}

func TestHashPasswordDifferentHashes(t *testing.T) {
	password := "MyPassword123"
	hash1, err := HashPassword(password)
	if err != nil {
		t.Fatalf("HashPassword() error = %v", err)
	}

	hash2, err := HashPassword(password)
	if err != nil {
		t.Fatalf("HashPassword() error = %v", err)
	}

	if hash1 == hash2 {
		t.Error("HashPassword() should generate different hashes for same password (due to salt)")
	}
}

func TestComparePassword(t *testing.T) {
	tests := []struct {
		name     string
		password string
		wantErr  bool
	}{
		{
			name:     "matching password",
			password: "MyPassword123",
			wantErr:  false,
		},
		{
			name:     "empty password",
			password: "",
			wantErr:  false,
		},
		{
			name:     "special characters",
			password: "P@ssw0rd!#$%",
			wantErr:  false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			hash, err := HashPassword(tt.password)
			if err != nil {
				t.Fatalf("HashPassword() error = %v", err)
			}

			if !ComparePassword(tt.password, hash) {
				t.Error("ComparePassword() should return true for matching password")
			}

			if ComparePassword(tt.password+"wrong", hash) {
				t.Error("ComparePassword() should return false for non-matching password")
			}

			if ComparePassword("", hash) && tt.password != "" {
				t.Error("ComparePassword() should return false for empty password when hash is not empty")
			}
		})
	}
}

func TestGenerateToken(t *testing.T) {
	tests := []struct {
		name     string
		userID   string
		username string
	}{
		{
			name:     "valid token",
			userID:   "user-123",
			username: "testuser",
		},
		{
			name:     "empty username",
			userID:   "user-456",
			username: "",
		},
		{
			name:     "special characters in username",
			userID:   "user-789",
			username: "test_user-123",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			token, err := GenerateToken(tt.userID, tt.username)
			if err != nil {
				t.Fatalf("GenerateToken() error = %v", err)
			}

			if token == "" {
				t.Error("GenerateToken() returned empty token")
			}

			// JWT should have 3 parts separated by dots
			parts := strings.Split(token, ".")
			if len(parts) != 3 {
				t.Errorf("GenerateToken() returned invalid JWT format, got %d parts, want 3", len(parts))
			}
		})
	}
}

func TestGenerateTokenDifferentUsers(t *testing.T) {
	token1, err := GenerateToken("user-1", "user1")
	if err != nil {
		t.Fatalf("GenerateToken() error = %v", err)
	}

	token2, err := GenerateToken("user-2", "user2")
	if err != nil {
		t.Fatalf("GenerateToken() error = %v", err)
	}

	if token1 == token2 {
		t.Error("GenerateToken() should generate different tokens for different users")
	}
}

func TestVerifyToken(t *testing.T) {
	tests := []struct {
		name      string
		userID    string
		username  string
		wantError bool
	}{
		{
			name:      "valid token",
			userID:    "user-123",
			username:  "testuser",
			wantError: false,
		},
		{
			name:      "empty username",
			userID:    "user-456",
			username:  "",
			wantError: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			token, err := GenerateToken(tt.userID, tt.username)
			if err != nil {
				t.Fatalf("GenerateToken() error = %v", err)
			}

			claims, err := VerifyToken(token)
			if (err != nil) != tt.wantError {
				t.Errorf("VerifyToken() error = %v, wantError %v", err, tt.wantError)
				return
			}

			if !tt.wantError {
				if claims.UserID != tt.userID {
					t.Errorf("VerifyToken() UserID = %v, want %v", claims.UserID, tt.userID)
				}
				if claims.Username != tt.username {
					t.Errorf("VerifyToken() Username = %v, want %v", claims.Username, tt.username)
				}
			}
		})
	}
}

func TestVerifyTokenInvalid(t *testing.T) {
	tests := []struct {
		name  string
		token string
	}{
		{
			name:  "invalid token",
			token: "invalid.token.here",
		},
		{
			name:  "malformed token",
			token: "not-a-jwt-token",
		},
		{
			name:  "empty token",
			token: "",
		},
		{
			name:  "random string",
			token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.invalid.signature",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := VerifyToken(tt.token)
			if err == nil {
				t.Error("VerifyToken() should return error for invalid token")
			}
			if !strings.Contains(err.Error(), "invalid or expired token") {
				t.Errorf("VerifyToken() error message = %v, want 'invalid or expired token'", err.Error())
			}
		})
	}
}

func TestVerifyTokenExpired(t *testing.T) {
	// Create a token that's already expired
	claims := JWTClaims{
		UserID:   "user-123",
		Username: "testuser",
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(-1 * time.Hour)), // Expired 1 hour ago
			IssuedAt:  jwt.NewNumericDate(time.Now().Add(-2 * time.Hour)),
			NotBefore: jwt.NewNumericDate(time.Now().Add(-2 * time.Hour)),
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	tokenString, err := token.SignedString(jwtSecret)
	if err != nil {
		t.Fatalf("Failed to create test token: %v", err)
	}

	_, err = VerifyToken(tokenString)
	if err == nil {
		t.Error("VerifyToken() should return error for expired token")
	}
}

func TestSetJWTSecret(t *testing.T) {
	originalSecret := jwtSecret
	defer func() {
		jwtSecret = originalSecret // Restore original secret
	}()

	newSecret := "new-test-secret-key"
	SetJWTSecret(newSecret)

	if string(jwtSecret) != newSecret {
		t.Errorf("SetJWTSecret() = %v, want %v", string(jwtSecret), newSecret)
	}

	// Verify tokens work with new secret
	token, err := GenerateToken("test-user", "testuser")
	if err != nil {
		t.Fatalf("GenerateToken() error = %v", err)
	}

	claims, err := VerifyToken(token)
	if err != nil {
		t.Fatalf("VerifyToken() error = %v", err)
	}

	if claims.UserID != "test-user" {
		t.Errorf("Token with new secret: UserID = %v, want test-user", claims.UserID)
	}
}

func TestIsValidEmail(t *testing.T) {
	tests := []struct {
		name  string
		email string
		want  bool
	}{
		// Valid emails
		{name: "standard email", email: "user@example.com", want: true},
		{name: "email with dot", email: "test.user@example.com", want: true},
		{name: "email with plus", email: "user+tag@example.co.uk", want: true},
		{name: "email with numbers", email: "user123@test-domain.com", want: true},
		{name: "short email", email: "a@b.co", want: true},
		{name: "subdomain", email: "user@mail.example.com", want: true},

		// Invalid emails
		{name: "no @", email: "invalid", want: false},
		{name: "no domain", email: "invalid@", want: false},
		{name: "no user", email: "@example.com", want: false},
		{name: "no TLD", email: "user@example", want: false},
		{name: "space in email", email: "user @example.com", want: false},
		{name: "space in domain", email: "user@exam ple.com", want: false},
		{name: "empty string", email: "", want: false},
		{name: "dot at start of domain", email: "user@.com", want: false},
		{name: "too long", email: strings.Repeat("a", 256) + "@example.com", want: false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := IsValidEmail(tt.email); got != tt.want {
				t.Errorf("IsValidEmail(%q) = %v, want %v", tt.email, got, tt.want)
			}
		})
	}
}

func TestIsValidUsername(t *testing.T) {
	tests := []struct {
		name     string
		username string
		want     bool
	}{
		// Valid usernames
		{name: "alphanumeric", username: "user123", want: true},
		{name: "with underscore", username: "test_user", want: true},
		{name: "mixed case", username: "Player_123", want: true},
		{name: "minimum length", username: "abc", want: true},
		{name: "maximum length", username: strings.Repeat("a", 50), want: true},
		{name: "all uppercase", username: "TESTUSER", want: true},
		{name: "all numbers", username: "123456", want: true},

		// Invalid usernames
		{name: "too short", username: "ab", want: false},
		{name: "too long", username: strings.Repeat("a", 51), want: false},
		{name: "with dash", username: "user-name", want: false},
		{name: "with dot", username: "user.name", want: false},
		{name: "with space", username: "user name", want: false},
		{name: "with @", username: "user@name", want: false},
		{name: "with #", username: "user#123", want: false},
		{name: "empty", username: "", want: false},
		{name: "special chars", username: "user!@#", want: false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := IsValidUsername(tt.username); got != tt.want {
				t.Errorf("IsValidUsername(%q) = %v, want %v", tt.username, got, tt.want)
			}
		})
	}
}

func TestIsValidPassword(t *testing.T) {
	tests := []struct {
		name     string
		password string
		want     bool
	}{
		// Valid passwords
		{name: "standard password", password: "Password123", want: true},
		{name: "with special chars", password: "MyP@ssw0rd", want: true},
		{name: "minimum valid", password: "Abcd1234", want: true},
		{name: "long password", password: "ThisIsAVeryLongPassword123", want: true},
		{name: "multiple special", password: "P@ss!w0rd#123", want: true},

		// Invalid passwords - missing uppercase
		{name: "no uppercase", password: "password123", want: false},
		{name: "all lowercase with number", password: "test1234", want: false},

		// Invalid passwords - missing lowercase
		{name: "no lowercase", password: "PASSWORD123", want: false},
		{name: "all uppercase with number", password: "TEST1234", want: false},

		// Invalid passwords - missing number
		{name: "no number", password: "Password", want: false},
		{name: "only letters", password: "TestPass", want: false},

		// Invalid passwords - too short
		{name: "too short", password: "Pass1", want: false},
		{name: "very short", password: "Ab1", want: false},
		{name: "empty", password: "", want: false},

		// Edge cases
		{name: "exactly 8 chars", password: "Pass1234", want: true},
		{name: "7 chars valid format", password: "Pass123", want: false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := IsValidPassword(tt.password); got != tt.want {
				t.Errorf("IsValidPassword(%q) = %v, want %v", tt.password, got, tt.want)
			}
		})
	}
}

func TestTokenPayload(t *testing.T) {
	userID := "test-user-id-123"
	username := "testuser"

	token, err := GenerateToken(userID, username)
	if err != nil {
		t.Fatalf("GenerateToken() error = %v", err)
	}

	// Decode without verification to check payload
	parser := jwt.NewParser()
	decoded, _, err := parser.ParseUnverified(token, &JWTClaims{})
	if err != nil {
		t.Fatalf("ParseUnverified() error = %v", err)
	}

	claims, ok := decoded.Claims.(*JWTClaims)
	if !ok {
		t.Fatal("Failed to cast claims to JWTClaims")
	}

	if claims.UserID != userID {
		t.Errorf("Token payload UserID = %v, want %v", claims.UserID, userID)
	}

	if claims.Username != username {
		t.Errorf("Token payload Username = %v, want %v", claims.Username, username)
	}

	if claims.ExpiresAt == nil {
		t.Error("Token should have ExpiresAt")
	}

	if claims.IssuedAt == nil {
		t.Error("Token should have IssuedAt")
	}
}

func BenchmarkHashPassword(b *testing.B) {
	password := "BenchmarkPassword123"
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, _ = HashPassword(password)
	}
}

func BenchmarkComparePassword(b *testing.B) {
	password := "BenchmarkPassword123"
	hash, _ := HashPassword(password)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = ComparePassword(password, hash)
	}
}

func BenchmarkGenerateToken(b *testing.B) {
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, _ = GenerateToken("user-123", "testuser")
	}
}

func BenchmarkVerifyToken(b *testing.B) {
	token, _ := GenerateToken("user-123", "testuser")
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, _ = VerifyToken(token)
	}
}
